rlp/src/rlpin.rs

```diff
@@ -342,43 +330,38 @@ impl<'a> BasicDecoder<'a> {
 	pub fn decode_value<T, F>(&self, f: F) -> Result<T, DecoderError>
 		where F: Fn(&[u8]) -> Result<T, DecoderError> {
 
-		let bytes = self.rlp.as_raw();
-
-		match bytes.first().cloned() {
-			// RLP is too short.
-			None => Err(DecoderError::RlpIsTooShort),
-			// Single byte value.
-			Some(l @ 0...0x7f) => Ok(f(&[l])?),
-			// 0-55 bytes
-			Some(l @ 0x80...0xb7) => {
-				let last_index_of = 1 + l as usize - 0x80;
-				if bytes.len() < last_index_of {
-					return Err(DecoderError::RlpInconsistentLengthAndData);
-				}
-				let d = &bytes[1..last_index_of];
-				if l == 0x81 && d[0] < 0x80 {
-					return Err(DecoderError::RlpInvalidIndirection);
-				}
-				Ok(f(d)?)
-			},
-			// Longer than 55 bytes.
-			Some(l @ 0xb8...0xbf) => {
-				let len_of_len = l as usize - 0xb7;
-				let begin_of_value = 1 as usize + len_of_len;
-				if bytes.len() < begin_of_value {
-					return Err(DecoderError::RlpInconsistentLengthAndData);
-				}
-				let len = decode_usize(&bytes[1..begin_of_value])?;
+		let bytes = self.rlp;
 
-				let last_index_of_value = begin_of_value.checked_add(len)
-					.ok_or(DecoderError::RlpInvalidLength)?;
-				if bytes.len() < last_index_of_value {
-					return Err(DecoderError::RlpInconsistentLengthAndData);
-				}
-				Ok(f(&bytes[begin_of_value..last_index_of_value])?)
+		let l = *bytes.first().ok_or_else(|| DecoderError::RlpIsTooShort)?;
```

Comment:
2nd optimization:

- do not call `cloned()`

Reply:
```cloned() == into_iter().map(|&x| x.clone)``` ?

hrm, have you checked the assembly from `cloned()`?  I would expect this to be optimized to just a single indirection since the address to `bytes` should be in a register but I'm not sure!

`deref` is better here but shouldn´t impact ```performance``` (should produce the same asm), however it is more clear because `clone` is considered an expensive operation

Reply:
I haven't looked at asm. it was just ~3% faster every time I run `bench_decode_1000_values` bench